frontend/apps/reader/modules/readertoc.lua

```diff
@@ -275,147 +280,174 @@ function ReaderToc:getTocTitleOfCurrentPage()
+A negative level does the same, but computes the depth level in reverse (i.e., -1 is the most deeply nested one).
+Omitting the level argument returns the full hierarchical table.
 --]]
 function ReaderToc:getTocTicks(level)
-    if self.ticks[level] then return self.ticks[level] end
-    -- build toc ticks if not found
+    -- Handle negative levels
+    if level and level < 0 then
+        level = self:getMaxDepth() + level + 1
+    end
+
+    if level then
+        if self.ticks and self.ticks[level] then
+            return self.ticks[level]
+        end
+    else
+        if self.ticks then
+            return self.ticks
+        end
+    end
+
+    -- Build ToC ticks if not found
     self:fillToc()
-    local ticks = {}
+    self.ticks = {}
 
     if #self.toc > 0 then
-        if level == 0 then
-            local depth = 0
-            for i = #self.toc, 1, -1 do
-                local v = self.toc[i]
-                if v.depth >= depth then
-                    table.insert(ticks, v.page)
-                end
-                depth = v.depth
-            end
-        else
-            local depth
-            if level > 0 then
-                depth = level
-            else
-                depth = self:getMaxDepth() + level + 1
-            end
-            for _, v in ipairs(self.toc) do
-                if v.depth == depth then
-                    table.insert(ticks, v.page)
-                end
+        -- Start by building a simple hierarchical ToC tick table
+        for _, v in ipairs(self.toc) do
+            if not self.ticks[v.depth] then
+                self.ticks[v.depth] = {}
             end
+            table.insert(self.ticks[v.depth], v.page)
+        end
+
+        -- Normally the ticks are already sorted, but in rare cases,
+        -- ToC nodes may be not in ascending order
+        for k, _ in ipairs(self.ticks) do
+            table.sort(self.ticks[k])
         end
-        -- normally the ticks are sorted already but in rare cases
-        -- toc nodes may be not in ascending order
-        table.sort(ticks)
-        -- cache ticks only if ticks are available
-        self.ticks[level] = ticks
     end
-    return ticks
+
+    if level then
+        return self.ticks[level]
+    else
+        return self.ticks
+    end
 end
 
-function ReaderToc:getTocTicksForFooter()
+--[[
+Returns a flattened list of ToC ticks, without duplicates
+]]
+function ReaderToc:getTocTicksFlattened()
+    if self.ticks_flattened then return self.ticks_flattened end
+
+    -- It hasn't been cached yet, compute it.
+    local ticks = self:getTocTicks()
     local ticks_candidates = {}
-    local max_level = self:getMaxDepth()
-    for i = 0, -max_level, -1 do
-        local ticks = self:getTocTicks(i)
-        table.insert(ticks_candidates, ticks)
+
+    for _, v in ipairs(ticks) do
+        for depth, page in ipairs(v) do
+            table.insert(ticks_candidates, page)
```

Comment:
You could do the duplicate/hash check here, and not insert it if page already seen. (And then, no need for the intermediate ticks_candidate.)